helix-term/src/ui/editor.rs

```diff
@@ -1205,16 +1224,9 @@ impl Component for EditorView {
                 if cx.editor.should_close() {
                     return EventResult::Ignored(None);
                 }
-                let config = cx.editor.config();
-                let (view, doc) = current!(cx.editor);
-                view.ensure_cursor_in_view(doc, config.scrolloff);
-
-                // Store a history state if not in insert mode. This also takes care of
-                // committing changes when leaving insert mode.
-                if doc.mode() != Mode::Insert {
-                    doc.append_changes_to_history(view.id);
-                }
+                self.complete_change(&mut cx);
```

Comment:
I think this code should not be abstracted between key events and paste: I think it would be better if the code were duplicated. We want to ensure the cursor is in view and save the pasted change to history when pasting in any mode, and we shouldn't need to execute any mode transition hooks. But this block is currently wrong for key events and needs some modification since it may use the wrong document and view when switching between views (https://github.com/helix-editor/helix/pull/3508).

Reply:
Makes sense to me on the separation between key events and paste. I've changed this to duplicate the code instead of sharing it.

I'm not sure I follow on "we want to [...] save the pasted change to history when pasting in any mode" though. To me, if I'm in insert and I hit my OS paste key, I expect that paste to be part of the ongoing insert. If I edit the inserted text from pasting and then exit insert mode, I expect undo to get the whole insert, not the paste and edit separately. Maybe I'm missing something though?

Reply:
Oh hmm yeah I'm not sure what I was thinking. Just when mode isn't insert we should be saving history 👍